subsys/net/ip/net_core.c

```diff
@@ -419,4 +419,11 @@ static int net_init(struct device *unused)
 	return status;
 }
 
+/* The networking subsystem requires the main thread to execute at a
+ * cooperative priority to function correctly. If this build assert triggers
+ * verify your configuration to ensure that cooperative threads are enabled
+ * and that the main thread priority is negative (cooperative).
+ */
+BUILD_ASSERT(CONFIG_MAIN_THREAD_PRIORITY < 0);
+
```

Comment:
Invalid configurations should be detected during configuration instead
of during compilation whenever possible.

I believe that adding the below at an appropriate net Kconfig location would achieve this.

```
# The networking subsystem requires the main thread to execute at
# a cooperative priority.
config MAIN_THREAD_PRIORITY
	range -256 -1
```

Reply:
I don't see that this is going to do much anyway.  The fact that the main thread is or is not preemptible says nothing about users of the network APIs which may or may not be running in the main thread.  Not, for example, that ztest always spawns a new thread for every test, so **none** of the existing tests would be caught by this check if they happened to be hitting the network out of a preemptible thread.

I'd suggest a runtime __ASSERT(k_thread_prioroity_get(k_current_get() < 0) at the relevant API entry points.

Reply:
> I believe that adding the below at an appropriate net Kconfig location would achieve this.
> config MAIN_THREAD_PRIORITY
>        range -256 -1

This will give error if the default MAIN_THREAD_PRIORITY is 0. In which case user needs to add the proper value into prj.conf file. This PR tries to avoid this.